### PR TITLE
Don't auto disable session affinity with kube-proxy

### PR DESCRIPTION
**Pull request by liuyuan10**
Right now, session affinity is being automatically disabled when
running with kube-proxy. This leaves pod with no session affinity
support.

Actually the session affinity is already supported in the lxc tc hook so
we can enable it without host reachable services.

Signed-off-by: Yuan Liu <liuyuan@google.com>

```release-note
Enable session affinity by default and avoid disabling it when used with kube-proxy
```

### CODE REVIEW

File: daemon/cmd/kube_proxy_replacement.go
Comment (borkmann) [inline daemon/cmd/kube_proxy_replacement.go:59]: Overall looks reasonable to me. I think we never considered extending this for the 'legacy' kube-proxy case, but makes sense to have this for ClusterIP there. The change would break with older kernels that do not have LRU map though, that still needs some minor tweaking in your PR, see ...
```
	if option.Config.EnableSessionAffinity {
		if !probesManager.GetMapTypes().HaveLruHashMapType {
			msg := "SessionAffinity feature requires BPF LRU maps"
			if strict {
				log.Fatal(msg)
			} else {
				log.Warnf("%s. Disabling the feature.", msg)
				option.Config.EnableSessionAffinity = false
			}
		}
	}
```
... where this would fatal given strict is set.

Comment (liuyuan10) [reply daemon/cmd/kube_proxy_replacement.go:59]: What behavior do you expect to have here? I kind of prefer the current way where it fatals when a feature enabled by users can't be turned on. Note this flag is false by default.

Comment (liuyuan10) [reply daemon/cmd/kube_proxy_replacement.go:59]: I can remove the log.Fatal if that's what you want.

Comment (brb) [reply daemon/cmd/kube_proxy_replacement.go:59]: Re log.Fatal: I think it's fine as it is, as running with `--kube-proxy-replacement=disabled` the `strict` is always false, so we would never fatal.

**Review by brb — COMMENTED** [review]
Changes LGTM as they are. One thing - have you considered making `--enable-session-affinity` to default to `true` considering that `--disable-k8s-services` defaults to `false`, which makes Cilium to handle ClusterIP svcs in non-host netns?

File: daemon/cmd/kube_proxy_replacement.go
Comment (liuyuan10) [reply daemon/cmd/kube_proxy_replacement.go:59]: When `--kube-proxy-replacement=disabled`, `strict` is also true:

```
strict = option.Config.KubeProxyReplacement != option.KubeProxyReplacementProbe
```

While I'd like to enforce this feature if people enable it specifically, it will cause crash for those running v4.9.17 - v4.10. We may want to remove this log.Fatal to avoid such crash. For this PR, I simply want people to have a way to enable it in kube-proxy mode.

**Comment by liuyuan10** [comment]
> Changes LGTM as they are. One thing - have you considered making `--enable-session-affinity` to default to `true` considering that `--disable-k8s-services` defaults to `false`, which makes Cilium to handle ClusterIP svcs in non-host netns?

I don't want to make it default true. As I mentioned above, it causes crash for those running v4.9.17 to v4.10.

File: daemon/cmd/kube_proxy_replacement.go
Comment (brb) [reply daemon/cmd/kube_proxy_replacement.go:59]: OK, I see. But anyway when `--kube-proxy-replacement=disabled`, then the init function will return before the strict checks (incl. the one for the LRU) happen: https://github.com/cilium/cilium/blob/master/daemon/cmd/kube_proxy_replacement.go#L68. So, the comment https://github.com/cilium/cilium/pull/13480/files#r502304131 still needs to be addressed then.

**Review by brb — CHANGES_REQUESTED** [review]
https://github.com/cilium/cilium/pull/13480/files#r502304131 needs to be addressed.

**Comment by brb** [comment]
> I don't want to make it default true. As I mentioned above, it causes crash for those running v4.9.17 to v4.10.

If you are planning to get rid of the `log.Fatal` case when `--kube-proxy-replacement=disabled`, then it's safe to enable the feature by default.

File: daemon/cmd/kube_proxy_replacement.go
Comment (liuyuan10) [reply daemon/cmd/kube_proxy_replacement.go:59]: Added a section to disable session-affinity when `--kube-proxy-replacement=disabled`

**Review by liuyuan10 — COMMENTED** [review]
> > I don't want to make it default true. As I mentioned above, it causes crash for those running v4.9.17 to v4.10.
> 
> If you are planning to get rid of the `log.Fatal` case when `--kube-proxy-replacement=disabled`, then it's safe to enable the feature by default.

flipped the default value of EnableSessionAffinity.

File: daemon/cmd/kube_proxy_replacement.go
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go:58]: Instead of duplicating the check, maybe you can move the existing check into a function with a `strict` param, and then invoke the function from here with `strict=false`?

**Review by brb — CHANGES_REQUESTED** [review]
One comment below, and also can you please add a test for it under the `Checks ClusterIP connectivity` context in the `test/k8sT/Services.go`?

**Comment by nebril** [comment]
test-upstream-k8s

**Comment by brb** [comment]
Hey @liuyuan10, are you still planning to address the review comments?

**Comment by liuyuan10** [comment]
Yes, I will. I'm not seeing this is urgent so I'm taking my time

File: daemon/cmd/kube_proxy_replacement.go
Comment (liuyuan10) [reply daemon/cmd/kube_proxy_replacement.go:58]: done

**Comment by liuyuan10** [comment]
> One comment below, and also can you please add a test for it under the `Checks ClusterIP connectivity` context in the `test/k8sT/Services.go`?

@brb I reused the existing testSessionAffinity and added a test. However, I can't install vagrant on my corp machines due to security policies so I don't know how to test it. Can the CI help me to run the test?

**Comment by brb** [comment]
@liuyuan10 Thanks for adding the tests. Please run `make -C Documentation update-cmdref` to update the cmdref help pages (this should fix the travis-ci build).

> Can the CI help me to run the test?

Yes, I will trigger the test run after you have updated the above. However, just noticed, that we only run with kube-proxy in the `K8s-1.19-kernel-4.9` job which kernel doesn't have the LRU feature required by session affinity. I will need to test it manually. Also, if the test doesn't fail on the CI, then we should keep it as it is.

**Review by christarazi — CHANGES_REQUESTED** [review]
LGTM for the CI code changes; I've only reviewed CI code.

Also, could we change the release note to avoid the double negative? Something like:

```
Enable session affinity by default and avoid disabling it when used with kube-proxy
```

I find that the above reads much easier.

**Comment by liuyuan10** [comment]
> Please run make -C Documentation update-cmdref to update the cmdref help pages (this should fix the travis-ci build).

Done

> if the test doesn't fail on the CI, then we should keep it as it is

@brb If the session affinity is disabled due to kernel version, certainly the new test will fail, won't it? Should I skip the test based on kernel version like this?
```
func RunsOn410Kernel()OrLater bool {
  return os.Getenv("KERNEL") >= "410"
}
```

**Comment by liuyuan10** [comment]
> Also, could we change the release note to avoid the double negative? Something like:

@christarazi Done

**Review by christarazi — APPROVED** [review]

**Comment by brb** [comment]
retest-4.9

**Comment by brb** [comment]
retest-gke

**Comment by liuyuan10** [comment]
I think the "K8s-1.19-kernel-4.9" is failing because session affinity is disabled...

**Comment by brb** [comment]
> I think the "K8s-1.19-kernel-4.9" is failing because session affinity is disabled...

Yeah, it's expected. I just wanted to double check that it fails indeed.

> @brb If the session affinity is disabled due to kernel version, certainly the new test will fail, won't it? Should I skip the test based on kernel version like this?

The approach SGTM. However, not in all cases we pass the KERNEL env param. So maybe you could determine from `uname -a` executed on any worker node?

**Comment by liuyuan10** [comment]
> The approach SGTM. However, not in all cases we pass the KERNEL env param. So maybe you could determine from uname -a executed on any worker node?

@brb Using uname -a requires extra work. I see other tests are simply using helpers.DoesNotRunOnNetNextOr419Kernel for LRU feature. Looks like we only run 419 kernels. I added a skip based on that

**Comment by brb** [comment]
> Using uname -a requires extra work. I see other tests are simply using helpers.DoesNotRunOnNetNextOr419Kernel for LRU feature. Looks like we only run 419 kernels. I added a skip based on that

Could you then specify in the comment that technically it could work on 4.10?

**Comment by liuyuan10** [comment]
> Could you then specify in the comment that technically it could work on 4.10?

Done

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
@liuyuan10 Thanks. Why are there so many changes in `Documentation/cmdref`?

**Comment by liuyuan10** [comment]
> @liuyuan10 Thanks. Why are there so many changes in `Documentation/cmdref`?

Those are generated by `make -C Documentation update-cmdref`. I get the same diffs from master. I moved them into a separate commit.

**Comment by brb** [comment]
> Those are generated by make -C Documentation update-cmdref. I get the same diffs from master. I moved them into a separate commit.

I took your `fixsa` branch, removed the https://github.com/cilium/cilium/pull/13480/commits/583bfdf21fc9fa0af3f388f61bf48b173b502f36 commit, and ran `make build && make -C Documentation update-cmdref`. It produced a single line change. Did you run `make build`? Also, after fixing this, please amend the changes to the main commit for the sake of git-bisect.

**Comment by liuyuan10** [comment]
@brb Thanks for helping with the issue. Done.

**Review by brb — APPROVED** [review]

**Comment by brb** [comment]
test-me-please

**Comment by joestringer** [comment]
Looks like the unit tests failed and it might be related to this PR:
```
=== RUN   Test

level=info msg="Connecting to etcd server..." config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore

level=info msg="Detected MTU 1460" subsys=mtu

level=error msg="Command execution failed" cmd="[bpftool -j feature probe]" error="exec: \"bpftool\": executable file not found in $PATH" subsys=probes

level=fatal msg="could not run bpftool" error="exec: \"bpftool\": executable file not found in $PATH" subsys=probes

level=info msg="Got lease ID 49ec761897ccee05" subsys=kvstore

level=info msg="Got lock lease ID 49ec761897ccee07" subsys=kvstore

level=info msg="Initial etcd session established" config= endpoints="[http://127.0.0.1:4002]" subsys=kvstore

level=info msg="Successfully verified version of etcd endpoint" config= endpoints="[http://127.0.0.1:4002]" etcdEndpoint="http://127.0.0.1:4002" subsys=kvstore version=3.3.20

FAIL	github.com/cilium/cilium/daemon/cmd	2.093s

FAIL

make: *** [Makefile:222: unit-tests] Error 1
```

https://travis-ci.com/github/cilium/cilium/jobs/450376083

Additionally, the runtime tests failed during inital deploy with an error that looks related:

```
02:24:56      runtime: Nov 30 10:24:53 runtime cilium-agent[15347]: level=fatal msg="SessionAffinity feature requires BPF LRU maps" subsys=daemon
02:24:56      runtime: Cilium failed to start
```
https://jenkins.cilium.io/job/Cilium-PR-Runtime-4.9/2785/execution/node/92/log/

**Comment by liuyuan10** [comment]
Not sure about the first failure. But for Runtime-4.9, I believe it's failing with a test case where KubeProxyReplacement = KubeProxyReplacementStrict. This is the only reason I can think of that triggers the log message but I can't see why we are setting KubeProxyReplacementStrict for 4.9. Switching the default value of enable-session-affinity to false should fix the failures. @brb what do you think?

**Comment by liuyuan10** [comment]
> FAIL	github.com/cilium/cilium/daemon/cmd	2.093s

This is failing because now we need the bpftool binary when KubeProxyReplacementDisabled. Previously it doesn't need it. The best is if we can have bpftool installed on the machine where the test is run. Not sure what's the best approach right now.

**Comment by liuyuan10** [comment]
I've fixed the Runtime-4.9 test by not enforcing SessionAffinity for partial kubeproxy mode. But as I said above, we need bpftool for unit test to run...

**Comment by joestringer** [comment]
> But as I said above, we need bpftool for unit test to run to fix the other test

@liuyuan10 ideally we would abstract out the implementation via interfaces + dependency injection so that the unit tests aren't relying directly on bpftool, it's mocked out. I would suggest trying to see if that's possible. If not, we have a fallback hack `option.Config.DryMode` which basically means "unit test mode" which logic can be switched on.

**Comment by brb** [comment]
@joestringer Alternative solution for this PR would be to bundle `bpftool` into a travis-ci image. Or there is no such thing as image in travis-ci?

**Comment by joestringer** [comment]
@brb I don't think there's such a thing, and I don't think it's a good idea to have the basic go unit testing workflow become dependent on random tooling in the developer environment, especially platform-specific tools like bpftool.

**Comment by brb** [comment]
@joestringer Makes sense. 

@liuyuan10 You can disable the session affinity here https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_test.go#L110. That would remove the need for bpftool. We are testing the changes in the CI anyway.

**Comment by liuyuan10** [comment]
@brb Thanks for the suggestion. Done

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
CI provisioning error:
```
k8s1-1.20: sudo: kubeadm: command not found
```

**Comment by brb** [comment]
retest-net-next

**Comment by brb** [comment]
test_only --k8s_version=1.20 --kernel_version=4.9

**Comment by brb** [comment]
retest-4.9

**Comment by aanm** [comment]
@brb the k8s 1.20 is still not merged yet. As far as testing goes, all required tests have passed. If you feel like the PR is ready to be merged please add the `ready-to-merge` label.

**Comment by brb** [comment]
retest-net-next

**Comment by brb** [comment]
retest-net-next

**Comment by liuyuan10** [comment]
@brb Is there anything blocking the merge?

**Comment by brb** [comment]
> @brb Is there anything blocking the merge?

@liuyuan10 We are still discussing it internally whether the increase in the BPF program complexity is acceptable. We will get to you back by the end of this week.

**Comment by brb** [comment]
@liuyuan10 Can you please rebase against the latest master?

**Review by borkmann — APPROVED** [review]
I'm okay with the patch, ack. One thing I'd like to avoid is backports to 1.8 and potentially 1.9 at least since they're out for quite some time and given we don't have much test exposure aside from CI there's potential risk of breakage from verifier complexity side on some of the older kernels. Having it in master allows for additional exposure time. Alternatively, we could leave the default as-is and only for new Helm deployments enable it automatically, but if we go for master-only route we'll see if there's any complaint before it lands in an actual release. Patch still needs rebase. Thanks!

**Comment by liuyuan10** [comment]
test/k8sT/Services.go has been refactored recently and I need to rework. will try to get to it this weekend.

**Comment by liuyuan10** [comment]
@brb Have rebased and reworked test/k8sT/Services.go. Please help to test it.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
@pchaigno Do you see this PR complicating https://github.com/cilium/cilium/issues/14726 even more?

**Comment by joestringer** [comment]
Marking as "draft" given that this PR needs a rebase. When you would like additional review, scroll to the bottom of the page and click the "Ready for review" button.

**Comment by brb** [comment]
@liuyuan10 I have revived your PR in https://github.com/cilium/cilium/pull/16179 (minus enabling SA by default).

Considering that we will make it possible to completely disable Cilium's E-W LB when KPR=disabled in v1.11, i.e. kube-proxy will be responsible for Pod2ClusterIP LB (https://github.com/cilium/cilium/issues/16197), there will be no need to enable SA by default. Thus, we can close this PR.

**Pull request closed without merging**
